File: binary tree/is_symmetric.py
'''
Checks if the binary tree is symmetric
Symmetric: left and right subtrees are mirror images of each other
'''
# Recursive solution
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def isSymmetric(self, root):
        """
        :type root: Optional[TreeNode]
        :rtype: bool
        """
        def helper(leftSubtree, rightSubtree):
            if not leftSubtree and not rightSubtree:
                return True

            if not leftSubtree or not rightSubtree:
                return False
            if leftSubtree.val != rightSubtree.val:
                return False
            return helper(leftSubtree.left, rightSubtree.right) and helper(leftSubtree.right, rightSubtree.left)
        # if tree is empty
        if not root:
            return True
        return helper(root.left, root.right)

# Iterative solution
class Solution(object):
    def isSymmetric(self, root):
        if not root:
            return True
        q1 = [root.left]
        q2 = [root.right]
        while q1 and q2:
            leftSubtree = q1.pop(0)
            rightSubtree = q2.pop(0)
            
            if not leftSubtree and not rightSubtree:
                continue
            if not leftSubtree or not rightSubtree:
                return False
            if leftSubtree.val != rightSubtree.val:
                return False
            q1.append(leftSubtree.left)
            q2.append(rightSubtree.right)
            q1.append(leftSubtree.right)
            q2.append(rightSubtree.left)
        return True
        
        

File: binary tree/is_symmetric.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
 #include <queue>
class Solution {
public:
    bool isSymmetric(TreeNode* root) {

        // Create two queues
        if (root == nullptr)
        {
            return true;
        }
        queue <TreeNode*> q1, q2;

        q1.push(root->left);
        q2.push(root->right);


        while(!q1.empty() && !q2.empty())
        {
            TreeNode* left_node = q1.front(); q1.pop();
            TreeNode* right_node = q2.front(); q2.pop();
       
         
            if (left_node == nullptr && right_node == nullptr)
            {
                continue;
            }
      
            if (left_node == nullptr || right_node == nullptr)
            {
                return false;
            }
     
            if (left_node->val != right_node->val) 
            {
                return false;
            }
            q1.push(left_node->left);
            q2.push(right_node->right);
            q1.push(left_node->right);
            q2.push(right_node->left);

        }

        return true;
    }
};
